### DISCUSSION TITLE
BFGS with constraints

### DISCUSSION THREAD

**Discussion by jecampagne** [jax-ml/jax#8072] [General] [answered]
Hi, 

I do not know if it is possible to perform with the current JAX lib something corresponding to the `scipy.optimize.fmin_l_bfgs_b` optimisation, ie. minimization with constraints on parameters, as well as computation of the Hessian at the minimum point found?

Thanks.

**Comment by shoyer** [comment, marked as answer]
JAXopt has support for constrained optimization: https://jaxopt.github.io/stable/constrained.html

**Reply by jecampagne** [reply]
Hi @shoyer thanks to point our jaxopt
Do you know where I can ask question? For instance how to perform the equivalent of `scipy` simple code
```python
def f(x):
    #(x - 1)^2 + (y - 2)^2
    return (x[0] - 1)**2 + (x[1] - 2)**2

def constraint(x):
    return np.atleast_1d(1.5 - ((x[0] + 1)**2 + x[1]**2))

minimize(f, np.array([0, 0]), bounds=((-5, 5), (-5, 5)),
         constraints={"fun": constraint, "type": "ineq"})
```

**Reply by Algue-Rythme** [reply]
There are [wrappers for Scipy](https://jaxopt.github.io/stable/_autosummary/jaxopt.ScipyBoundedMinimize.html)  in JaxOpt, unfortunately they do not support `constraints` argument, only box constraints. Adding more complex constraints is a TODO.

You won't get any speed up though, because the underlying implementation calls Scipy. However, you can evaluate the hessian of your objective function at the minimum, using Autodiff:

```python
import jax.numpy as jnp
from jax import jit, jacfwd, jacrev
from jaxopt import ScipyBoundedMinimize

# I replaced tuple by array
def f(x):
    return jnp.sum((x - jnp.array([1, 2])) ** 2)

def hessian(fun):
  return jit(jacfwd(jacrev(fun)))

scipy_wrap = ScipyBoundedMinimize(fun=f, tol=1e-3)
x_init = jnp.array([0., 0.])
bounds= (jnp.array([-5, -5]), jnp.array([5, 5]))
sol, state = scipy_wrap.run(x_init, bounds=bounds)
hes = hessian(f)(sol)
print(sol, hes)
```

Here I used `jaxopt.ScipyBoundedMinimize` for generality but your function `f` fall under the framework of least square minimization so I could have used `jaxopt.ScipyLeastSquares`. Thanks to Autodiff and implicit differentiation, if your function `f` had another parameter you could have computed the gradient wrt that:

```python
import jax.numpy as jnp
from jax import jit, jacfwd, jacrev, grad
from jaxopt import ScipyBoundedMinimize

def f(x, theta):
    return jnp.sum((x - theta) ** 2)

scipy_wrap = ScipyBoundedMinimize(fun=f, tol=1e-3)
x_init = jnp.array([0., 0.])
bounds= (jnp.array([-5, -5]), jnp.array([5, 5]))
theta = jnp.array([1., 2.])
def run_wrapper(theta):
  sol, state = scipy_wrap.run(x_init, bounds, theta)
  return sol
g = jacrev(run_wrapper)(theta)
print(g) # print identity function since the optimum is always equal to theta here
```

Now, if you don't care about Scipy, and all you want is to solve your optimization problem, you can take a look at other solvers in the library.  I would suggest to use [jaxopt.ProjectedGradient](https://jaxopt.github.io/stable/_autosummary/jaxopt.ProjectedGradient.html#jaxopt.ProjectedGradient) and to provide your own projection function:

```python
from jaxopt.projection import projection_l2_ball
from jaxopt import ProjectedGradient
from jaxopt.projection import projection_box, projection_l2_ball

def f(x, theta):
    return jnp.sum((x - theta) ** 2)

def myproj(x, hyper_params):
  bounds, radius = hyper_params
  # or something like that, you'll have to check I'm not sure
  x = projection_l2_ball(x + jnp.array([1,0]), radius) - jnp.array([1,0])
  x = projection_box(x, bounds)
  return x

scipy_wrap = ProjectedGradient(fun=f, projection=myproj, tol=1e-3, maxiter=100)
x_init = jnp.array([0., 0.])
bounds= (jnp.array([-5, -5]), jnp.array([5, 5]))
radius = 1.5
theta = jnp.array([1., 2.])
hyper_params = bounds, radius
sol, state = scipy_wrap.run(x_init, hyper_params, theta)
print(sol)
```

If you are unhappy with precision or runtime, take a look at parameters in constructor of ProjectedGradient. Once again, you can compute the hessian or the derivative wrt any parameter of interest if you need to, because every solver in Jaxopt support implicit differentiation.

**Reply by jecampagne** [reply]
Ho Thanks.
```python
  x = projection_l2_ball(x + jnp.array([1,0]), radius) - jnp.array([1,0])
  x = projection_box(x, bounds)
```
I am not sure to understand exactly what these lines are doing here `x`is a 2D array. If you write them in mathematical language it would help me. Thanks

**Reply by Algue-Rythme** [reply]
In my example x is a 1D array of size 2.

You will find the maths here:

[projection_box](https://jaxopt.github.io/stable/_autosummary/jaxopt.projection.projection_box.html#jaxopt.projection.projection_box)
[projection_l2_ball](https://jaxopt.github.io/stable/_autosummary/jaxopt.projection.projection_l2_ball.html#jaxopt.projection.projection_l2_ball)

The function `projection_box` projects the elements within the bounds, so that <a href="https://www.codecogs.com/eqnedit.php?latex=-5\leq&space;x_i&space;\leq&space;5" target="_blank"><img src="https://latex.codecogs.com/svg.latex?-5\leq&space;x_i&space;\leq&space;5" title="-5\leq x_i \leq 5" /></a>

The function `projection_l2_ball` projects your x within the ball of center (0,0) and radius given in hyper-parameter. If my understanding of your code snippet is correct, you want to constraint x within the ball of center (1,0) and radius sqrt(1.5) (yep, I did a mistake in my code I forgot sqrt). So in order to use `projection_l2_ball` I translate `x`, and now that I think about it I also made a mistake with the sign: `projection_l2_ball(x - jnp.array([1,0]), radius) + jnp.array([1,0])`. Once translated, the constraint can be reformulated as a projection within a ball of center (0,0) so we can use `projection_l2_ball`. And finally, I do the translation in opposite direction to ensure everything is in the right place.  

Note that both box constraints and l2 balls are convex set, hence their intersection is once again a convex set, and in this case Projected gradient descent is guaranteed to reach their intersection (if your objective is convex too, and smooth, which is the case here) after enough steps. So after a _sufficiently high_ number of iterations, you should expect to find the minimum of `f` under the constraints you chose.

**Reply by jecampagne** [reply]
Well, well here is snippet:
```python
def myproj(x, hyper_params):
    bounds, radius = hyper_params
    x = projection_l2_ball(x + jnp.array([1,0]), radius) - jnp.array([1,0])
    return x

bounds= (jnp.array([-10., -10.]), jnp.array([10., 10.]))
radius = 5.**2
hyper_params = bounds, radius

rng_key = jax.random.PRNGKey(42)
x = jax.random.uniform(rng_key, (1000,2), minval=-5,maxval=5)
p_x=myproj(x,hyper_params)

fig, ax = plt.subplots(figsize=(5,5))
ax.scatter(x[:,0],x[:,1])
ax.scatter(p_x[:,0],p_x[:,1])
```
Here is the result:

![image](https://user-images.githubusercontent.com/20539759/136181165-37aa517e-601c-4686-90f4-ad95c67326dd.png)


It is not really what I was expecting that is to say a disk of radius 5 centered on (-1,0). No?
`

**Reply by Algue-Rythme** [reply]
Oh yes, I should have precised: all Jaxopt functions and optimizers are meant to be applied on single example, not whole batch.
Thanks to `jax.vmap` this is not an issue since we can wrap `.run()` with it and optimize over batchs.  
This code works:

```python
from jaxopt.projection import projection_l2_ball
import jax
import matplotlib.pyplot as plt

def myproj(x, hyper_params):
    bounds, radius = hyper_params
    x = projection_l2_ball(x - jnp.array([1,0]), radius) + jnp.array([1,0])
    return x

bounds= (jnp.array([-10., -10.]), jnp.array([10., 10.]))
radius = 1.5
hyper_params = bounds, radius

rng_key = jax.random.PRNGKey(42)
x = jax.random.uniform(rng_key, (1000,2), minval=-5,maxval=5)
p_x=jax.vmap(myproj, in_axes=(0,None))(x,hyper_params)

fig, ax = plt.subplots()
ax.scatter(x[:,0],x[:,1])
ax.scatter(p_x[:,0],p_x[:,1])
```

If you are used to Numpy this might be surprising at first glance. However it some sense it is closer from the "mathematical approach": you think and write your code for a single example, and let `vmap` handle the rest.

**Reply by jecampagne** [reply]
Great! 

here it is if `radius = 1.5**2`

![image](https://user-images.githubusercontent.com/20539759/136191434-a5e9bd4a-bcd8-48db-89e8-534b6107141b.png)


Now I was in trouble with the projection_box, but I found :)
```python
def myproj(x, hyper_params):
    bounds, radius = hyper_params
    x = projection_box(x+ jnp.array([1,1]), bounds) - jnp.array([1,1])
    return x

bounds= (jnp.array([-1, -1]), jnp.array([1., 1.]))
```

I get
![image](https://user-images.githubusercontent.com/20539759/136196174-a3127801-0f88-4050-80f2-4b809a70d786.png)

which is ok !

And here is a nice combinaison
![image](https://user-images.githubusercontent.com/20539759/136197223-a1f6a6cc-fc52-4276-97a7-d78adcef7255.png)


Bye the way, I'm always in trouble with in_axes of jax.vmap, do you have an hint to get right? 

Is there a way to define a poly line shape  (in 2D) of the projection set? 

Anyway Thanks for your clear explanations. I hope that the other kinds of constraints (ie. those available by scipy) will be implemented.

**Reply by Algue-Rythme** [reply]
Nice figures @jecampagne !
I guess you could define a poly line shape manually with your own tools, but not with tools in Jaxopt currently.

**Reply by Algue-Rythme** [reply]
For `vmap` I pass a tuple of length 2 to `in_axes` because the function `myproj` takes two arguments. I want to batch wrt the first argument, but not wrt the second, so I use `(0, None)`: as a result I project 1000 examples onto the same ball.

If I wanted a different ball with different radius for each x, I would use `in_axes=(0,0)` and I could pass an array of 1000 different radius (one for each x). Or I could use `in_axes=(None,0)` to have a single example projected onto 1000 different balls. 

Or I could decide to create x as `x = jax.random.uniform(rng_key, (2,1000), minval=-5,maxval=5)` instead (notice that rows and columns are now permuted) so I would have used `in_axes=(1,None)` because the batch size woul be in axis 1, not 0.

**Reply by jecampagne** [reply]
Good ! Thanks a lot @Algue-Rythme. We can close the thread. I do not know if there is a jaxopt tag/forum for discussing further if needed.

**Comment by jecampagne** [comment]
Hi @Algue-Rythme 
You mention `jaxopt.ScipyLeastSquares`but I do not manage to get the API in the doc ???

**Reply by Algue-Rythme** [reply]
It's not referenced in the doc yet.

You will have to read the source and the docstring: https://github.com/google/jaxopt/blob/782b0a918695b4c9f75717a3bb4e371b4884d98a/jaxopt/_src/scipy_wrappers.py#L605 good luck!

**Comment by jecampagne** [comment]
@Algue-Rythme have you an example of `jaxopt.projection.projection_polyhedron` in 2D, thanks.

Note also that in the doc
![image](https://user-images.githubusercontent.com/20539759/136352888-21e54774-af64-4fad-83da-8bf09fcc5d00.png)
the `w` is certainly related to the `params` (ie. general L2 norm constraint, moreover I would have expected this use-case in the "minimimzation with constraint solved by Lagrangian method... it is a matter of taste), in any case this is not clear enough.

**Reply by Algue-Rythme** [reply]
Thanks for the typo, this is indeed `params` and not `w`!

You will find examples of usage in tests/ folder: https://github.com/google/jaxopt/blob/03d5d3f6ebefbab3f1986a715789071580fb03c9/tests/projection_test.py#L279
(in 5D) in addition to the doc. Let me know if it works!

> I would have expected this use-case in the "minimimzation with constraint solved by Lagrangian method... it is a matter of taste  
  
L2 regularization is very common in ML (one of the major target of Jaxopt) and is rarely formulated as a hard constraint in this context (here Ridge regression).
Consequently there is no need to introduce the Lagrangian here, even though the link is direct. `l2reg` is better understood as an hyper-parameter of the algorithm that we can tune/optimize (and that ML practitioners will recognize) rather than a Lagrangian multiplier of some constrained optimization problem.
